### ISSUE TITLE
Feature Request:  Support Anthropic Claude

### ISSUE THREAD

**Issue by rosstimson** [rksm/org-ai#107]
A lot of folk are starting to switch to [Anthropic Claude](https://www.anthropic.com/claude) as it often gives better results than ChatGPT, I wonder if it would be possible to support it.

Many thanks for your work.

**Comment by ul** [comment]
Some notes from my attempt to hack the support locally in case anyone interested to give it a go.

Changes in my config:

```elisp
(setq org-ai-default-chat-model "claude-3-opus-20240229")
(setq org-ai-openai-chat-endpoint "https://api.anthropic.com/v1/messages")
(setq org-ai-default-max-tokens 4096) ;; required in Anthropic API 
(setq org-ai-openai-api-token anthropic-api-key)
```

Changes in `org-ai-openai.el`:

```diff
@@ -168,11 +168,9 @@ whether messages are provided."
 (defun org-ai--get-headers ()
   "Determine the correct headers based on the service."
   `(("Content-Type" . "application/json")
-    ,(cond
-      ((eq org-ai-service 'azure-openai)
-       `("api-key" . ,(org-ai--openai-get-token)))
-      (t
-       `("Authorization" . ,(encode-coding-string (string-join `("Bearer" ,(org-ai--openai-get-token)) " ") 'utf-8))))))
+    ("x-api-key" . ,(org-ai--openai-get-token))
+    ("anthropic-version" . "2023-06-01")
+    ))
 
 (defvar org-ai--current-request-buffer-for-stream nil
   "Internal var that stores the current request buffer.
@@ -293,6 +291,7 @@ from the OpenAI API."
 When `RESPONSE' is nil, it means we are done. `CONTEXT' is the
 context of the special block. `BUFFER' is the buffer to insert
 the response into."
+  ;; (message "RESPONSE %s" response)
   (if response
 
       ;; process response
@@ -311,13 +310,13 @@ the response into."
                 (backward-char))
 
               ;; insert text
-              (if-let* ((choices (or (alist-get 'choices response)
-                                     (plist-get response 'choices)))
-                        (choice (and (arrayp choices) (> (length choices) 0) (aref choices 0)))
-                        (delta (plist-get choice 'delta)))
+              ;; RESPONSE ('type message_start 'message ('id msg_01Mgig2iFaFinGwRGvAAYSFJ 'type message 'role assistant 'content [] 'model claude-3-opus-20240229 'stop_reason nil 'stop_sequence nil 'usage ('input_tokens 11 'output_tokens 1)))
+              ;; RESPONSE ('type content_block_delta 'index 0 'delta ('type text_delta 'text Hello))
+              ;; RESPONSE ('type message_stop)
+              (if-let* ((type (plist-get response 'type)))
                   (cond
-                   ((plist-get delta 'role)
-                    (let ((role (plist-get delta 'role)))
+                   ((string= type "message_start")
+                    (let ((role (plist-get (plist-get response 'message) 'role)))
                       (progn
                         (setq org-ai--current-chat-role role)
                         (cond
@@ -328,15 +327,24 @@ the response into."
                          ((string= role "system")
                           (insert "\n[SYS]: ")))
                         (run-hook-with-args 'org-ai-after-chat-insertion-hook 'role role))))
-                   ((plist-get delta 'content)
-                    (let ((text (plist-get delta 'content)))
+                   ((string= type "content_block_delta")
+                    (let ((text (plist-get (plist-get response 'delta) 'text)))
                       (when (or org-ai--chat-got-first-response (not (string= (string-trim text) "")))
                         (when (and (not org-ai--chat-got-first-response) (string-prefix-p "```" text))
                           ;; start markdown codeblock responses on their own line
                           (insert "\n"))
                         (insert (decode-coding-string text 'utf-8))
                         (run-hook-with-args 'org-ai-after-chat-insertion-hook 'text text))
-                      (setq org-ai--chat-got-first-response t)))))
+                      (setq org-ai--chat-got-first-response t)))
+                   ((string= type "message_stop")
+                    (with-current-buffer buffer
+                      (when org-ai--current-insert-position-marker
+                        (goto-char org-ai--current-insert-position-marker))
+                      (let ((text "\n\n[ME]: "))
+                        (insert text)
+                        (run-hook-with-args 'org-ai-after-chat-insertion-hook 'end text))
+                      (org-element-cache-reset))
+                    )))
```

**Comment by ul** [comment]
Yes, except that it doesn’t support system prompts, as mentioned. We need to lift system prompt from messages to the top level of the request.
I also hit a “Multibyte text in HTTP request” error when using, say, Cyrillic characters despite the payload wrapped into `encode-coding-string` in `org-ai--payload` but I did have time to investigate it properly yet.
Otherwise, was happily using it as a replacement for GPT-4 for a few days already, and I don’t think I’m going back.

**Comment by ul** [comment]
Okay, "solved" the multibyte mystery. For whatever reason as soon as the body gets multibyte characters (that are replaced when processed by `encode-coding-string`), we need to encode the token as well:

```elisp
  (defun org-ai-claude--get-headers ()
    `(("Content-Type" . "application/json")
      ("x-api-key" . ,(encode-coding-string (org-ai--openai-get-token) 'utf-8))
      ("anthropic-version" . "2023-06-01")))

  (advice-add 'org-ai--get-headers :override #'org-ai-claude--get-headers)
```

The token doesn't seem to be changed by this call (it consists of ASCII characters after all) but encoding headers seems to be a typical workaround for this problem based on what I was able to google. If anyone got an insight into Emacs/`url` guts and can explain what's going on here, I'd be grateful.

**Comment by rksm** [comment]
Support for the Anthropic Claude API is implemented as of version 0.4.4.

For usage please see the readme: https://github.com/rksm/org-ai?tab=readme-ov-file#anthropic--claude
